### PR TITLE
Implement `'a or_null`

### PR DESCRIPTION
**Pull request by dkalinichenko-js**
Do not update the jkind in `update_decl_jkind` if the manifest is available; always copy the manifest's jkind in `check_coherence`. This allows us to give types like `'a or_null` non-standard jkinds for their kinds (i.e. `value_or_null` for `Type_variant`).

Pre-define `or_null` when `Layouts` has level `Alpha`. Add tests for `or_null`.

`or_null` will be added to `stdlib_alpha` in the next PR.

### CODE REVIEW

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+1 | type nested = int or_null or_null
+                  ^^^^^^^^^^^
+Error: This type int or_null should be an instance of type ('a : value)
+       The layout of int or_null is value_or_null, because
+         it is the primitive value_or_null type or_null.
+       But the layout of int or_null must be a sublayout of value, because
+         the type argument of or_null has layout value.
+|}]
+
+let should_fail = This (This 5)
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:110]: Also worth testing: `This Null`.

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+       The layout of 'a t is value_or_null, because
+         it is the primitive value_or_null type or_null.
+       But the layout of 'a t must be a sublayout of value, because
+         of the definition of t at lines 1-3, characters 0-14.
+|}]
+
+let mk' n = `Foo (This n)
+
+[%%expect{|
+val mk' : 'a -> [> `Foo of 'a t ] = <fun>
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:127]: Again, I would expect an explicit kind on `'a`.

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+external this : 'a -> 'a or_null = "%identity"
+|}]
+
+external unsafe_get : 'a or_null -> 'a = "%identity"
+
+[%%expect{|
+external unsafe_get : 'a or_null -> 'a = "%identity"
+|}]
+
+let should_fail = [| Null; This 5 |]
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:175]: Probably should test this in a list, too.

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+Line 1, characters 19-32:
+1 | type should_fail = float or_null array
+                       ^^^^^^^^^^^^^
+Error: This type float or_null should be an instance of type
+         ('a : any_non_null)
+       The layout of float or_null is value_or_null, because
+         it is the primitive value_or_null type or_null.
+       But the layout of float or_null must be a sublayout of any_non_null, because
+         it's the type argument to the array type.
+|}]
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:201]: Maybe add a test with an `or_null` as a instance variable or a method in an object? I don't really care what the behavior is, but I wouldn't want a panic, say.

File: ocaml/typing/predef.ml [hunk 5]
```diff
@@ -460,6 +472,19 @@ let add_small_number_extension_types add_type env =
        ~jkind:(Jkind.Primitive.float32 ~why:(Primitive ident_unboxed_float32))
        ~jkind_annotation:Jkind.Const.Primitive.float32
 
+let add_or_null add_type env =
+  let add_type1 = mk_add_type1 add_type in
+  env
+  |> add_type1 ident_or_null
+  ~variance:Variance.covariant
+  ~separability:Separability.Sep
```
Comment (goldfirere) [inline ocaml/typing/predef.ml:480]: But this is a lie: this type is *not* separable. We've been thinking this is OK because we prevent `or_null`s from going into arrays. And I can't think of a problem here. But there should be a loud comment (and perhaps some discussion) about this.

Maybe, really, in our new world, we have the property that a type is separable if the type system thinks it is separable according to the old analysis _and_ the type is `non_null`. That's probably right.

File: ocaml/typing/typedecl.ml [hunk 6]
```diff
@@ -1610,6 +1607,17 @@ let update_decl_jkind env dpath decl =
       type_jkind
   in
 
+  (* Do not override the jkind if the manifest jkind if available. This is
```
Comment (goldfirere) [inline ocaml/typing/typedecl.ml:1610]:
```suggestion
  (* Do not override the jkind if the manifest jkind is available. This is
```

File: ocaml/typing/typedecl.ml [hunk 7]
```diff
@@ -1610,6 +1607,17 @@ let update_decl_jkind env dpath decl =
...
+     necessary to support types like ['a or_null], which have a non-standard
+     jkind. [check_coherence] will update the jkind to match the manifest. *)
```
Comment (goldfirere) [inline ocaml/typing/typedecl.ml:1612]: I don't see how this comment corresponds to the implementation. It says not to do something if there is a manifest.... but the code does something if there is a manifest. Could you try rewording this?

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+type 'a t = 'a or_null = Null | This of 'a
+|}]
+
+let to_option (x : 'a or_null) =
+  match x with
+  | Null -> None
+  | This x -> Some x
+
+[%%expect{|
+val to_option : 'a or_null -> 'a option = <fun>
```
[the remark this answers is not in the document]
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:20]: Oops. My mistake. Ignore.

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml (hunk 2 above)
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:127]: Oops. My mistake. Ignore.

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml (hunk 3 above)
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:189]: Done.

File: ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml (hunk 4 above)
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/test_or_null.ml:276]: Done.

File: ocaml/typing/predef.ml (hunk 5 above)
Comment (dkalinichenko-js) [reply ocaml/typing/predef.ml]: Added a comment.

File: ocaml/typing/typedecl.ml (hunk 7 above)
Comment (dkalinichenko-js) [reply ocaml/typing/typedecl.ml]: Edited.

**Review by goldfirere — APPROVED** [review]
